Support binary_sensor and device_tracker for HomeKit

Yonsm opened the pull request:
## Description:

Support binary_sensor and device_tracker for HomeKit.

Map binary_sensor and device_tracker to HomeKit BinarySensor, based on
‘device_class’  attributes:


    'gas': 'CarbonMonoxideSensor'
    'co2': 'CarbonDioxideSensor'
    'occupancy': 'OccupancySensor'
    'opening': 'ContactSensor'
    'motion': 'MotionSensor'
    'moisture': 'LeakSensor'
    'smoke': 'SmokeSensor'

If there is no any matched class (e.g. device_tracker), the default sensor type is OccupancySensor.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`.

Commit:
Merge remote-tracking branch 'home-assistant/dev' into dev

Commit:
Support binary_sensor and device_tracker for HomeKit

Map binary_sensor and device_tracker to HomeKit BinarySensor, based on
‘homekit_device_class’ or ‘device_class’  attributes:

'gas': 'CarbonMonoxideSensor'
'co2': 'CarbonDioxideSensor'
'occupancy': 'OccupancySensor'
'opening': 'ContactSensor'
'motion': 'MotionSensor'
'moisture': 'LeakSensor'
'smoke': 'SmokeSensor'

Default device_class is 'occupancy'.

Commit:
Support binary_sensor and device_tracker for HomeKit

Map binary_sensor and device_tracker to HomeKit BinarySensor, based on
‘homekit_device_class’ or ‘device_class’  attributes:

'gas': 'CarbonMonoxideSensor'
'co2': 'CarbonDioxideSensor'
'occupancy': 'OccupancySensor'
'opening': 'ContactSensor'
'motion': 'MotionSensor'
'moisture': 'LeakSensor'
'smoke': 'SmokeSensor'

Default device_class is 'occupancy'.

homeassistant commented:
Hi @Yonsm, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#13701).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Line wrap to 79 chars

Commit:
Line wrap to 79 chars

Yonsm commented:
This is my first pull request. And if success, I'll update more HomeKit support for:
CarbonDioxideSensor
LightSensor
AirQualitySensor
AirPurifier (fully support by HomeKit)

Commit:
Check 'home' for device_tracker for HomeKit

Commit:
Check 'home' for device_tracker for HomeKit

cdce8p commented on homeassistant/components/homekit/__init__.py:
```diff
@@ -92,6 +92,10 @@ def get_accessory(hass, state, aid, config):
 
+    elif state.domain == 'binary_sensor' or state.domain == 'device_tracker':
+        _LOGGER.debug('Add "%s" as "%s"', state.entity_id, 'BinarySensor')
+        return TYPES['BinarySensor'](hass, state, aid=aid)
```
You don't need to pass in the state, only the entity_id: `state.entity_id`.

Yonsm replied:
Only the entity_id is not enough, we need state.attributes to check binary sensor type. I'll follow your style to pass name, entity_id and one more param state.attributes.

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +75,46 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+class BinarySensor(HomeAccessory):
+    """Generate a BinarySensor accessory as binary sensor."""
+
+    def __init__(self, hass, state, **kwargs):
```
`entity_id` instead of `state`

Yonsm replied:
Fixed

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +75,46 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+
+    def __init__(self, hass, state, **kwargs):
+        """Initialize a BinarySensor accessory object."""
+        entity_id = state.entity_id
```
Isn't necessary if `entity_id` is passed to constructor.

Yonsm replied:
Fixed

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +75,46 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+            'opening': ('ContactSensor', 'ContactSensorState'),
+            'motion': ('MotionSensor', 'MotionDetected'),
+            'moisture': ('LeakSensor', 'LeakDetected'),
+            'smoke': ('SmokeSensor', 'SmokeDetected')}
```
Those should be constants defined in `./const.py`. E.g.:
 - `DEVICE_CLASS_GAS = 'gas'`
 - `SERV_GAS_SENSOR = CarbonMonoxideSensor`
 - `CHAR_GAS_DETECTED = CarbonMonoxideDetected`
 - ...

`service_map` a constant in `type_sensors.py` outside any class, like so:
```py
SERVICE_MAP = {
    DEVICE_CLASS_GAS: {SERV_GAS_SENSOR, CHAR_GAS_DETECTED),
    # ...
}
```

Yonsm replied:
Good coding style. I'll follow it.

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +75,46 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+
+        device_class_key = 'homekit_device_class' \
+            if ('homekit_device_class' in state.attributes) \
+            else 'device_class'
```
Currently there is no `homekit_device_class` attribute. For this PR we should concentrate just on the `device_class`.

Yonsm replied:
homekit_device_class it just like homebridge_outlet_type, etc in HomeBridge/homegbridge+homeassistant. It's preserved for user to assign this param in customization.yaml manually. Because sometimes there is no device_class in binary_sensor's attributes, and user can assign it to gas co/co2, or any other binary_sensor.

There will be more homekit_xxxx key for CO2/Light/AirQuality sensor to get correct sensor type instead of guessing by unit or entity id.  I'll document it in PR comment.

cdce8p replied:
We shouldn't write those in the state machine. I would prefer to leave it at `device_class` only for the moment. If you really want to implement this, use the `entity_config` dictionary from the `homekit` component. A good example is the `alarm_code` for `type_security_systems.py`.

Yonsm replied:
Remove it at this time.

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +75,46 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+        device_class_key = 'homekit_device_class' \
+            if ('homekit_device_class' in state.attributes) \
+            else 'device_class'
+        device_class = state.attributes.get(device_class_key)
```
Use `ATTR_DEVICE_CLASS` imported from `homeassistant.const` as key

Yonsm replied:
Just follow it

cdce8p reviewed (changes requested):
Great first Home Assistant PR! Looking forward for more to come.
I've left some comments.

Commit:
Refine params/const based on advice:

Commit:
Refine params/const based on advice:

Commit:
Trim trailing whitespace and align hanging indent

Commit:
Trim trailing whitespace and align hanging indent

Commit:
Trim trailing whitespace

Commit:
Trim trailing whitespace

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +84,53 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+class BinarySensor(HomeAccessory):
+    """Generate a BinarySensor accessory as binary sensor."""
+
+    def __init__(self, hass, entity_id, name, attributes, **kwargs):
```
Don't pass `attributes`. Use `hass` instead, like so:
`self.hass.states.get(self.entity_id).attributes`

Yonsm replied:
Fixed

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +84,53 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+        self.hass = hass
+        self.entity_id = entity_id
+
+        service_map = {
```
Please move this between `_LOGGER` and `@TYPES.register('TemperatureSensor')` and rename it to something like `BINARY_SENSOR_SERVICE_MAP`

Yonsm replied:
Fixed

cdce8p commented on homeassistant/components/homekit/const.py:
```diff
@@ -67,6 +73,26 @@
+DEVICE_CLASS_SMOKE = 'smoke'
+
+# #### Attributes ####
+ATTR_DEVICE_CLASS = 'device_class'
```
`ATTR_DEVICE_CLASS` is already defined in `homeassistant.const`

cdce8p commented on homeassistant/components/homekit/const.py:
```diff
@@ -43,7 +43,13 @@
+SERV_CONTACT_SENSOR = 'ContactSensor'
+SERV_MOTION_SENSOR = 'MotionSensor'
+SERV_LEAK_SENSOR = 'LeakSensor'
+SERV_SMOKE_SENSOR = 'SmokeSensor'
```
Please sort them alphabetically. The chars and device_classes as well.

Yonsm replied:
Follow it.
Furthermore, I also sort CHAR_CURRENT_HUMIDITY  and CHAR_CURRENT_POSITION, and this cause a conflict. Should I revert it, or would you resolve it?

cdce8p replied:
Try rebaseing it against `dev`: https://www.home-assistant.io/developers/development_catching_up

cdce8p reviewed (changes requested):
Please take a look at: https://github.com/home-assistant/home-assistant/pull/13701#discussion_r179712916 as well if you haven't seen it already.

Commit:
Revise param/const/device_class_key based on advice

1. Remove state param;
2. Sort const;
3. Import ATTR_DEVICE_CLASS from homeassistant.const;
4. Remove device_class_key:
       device_class_key = ATTR_HOMEKIT_DEVICE_CLASS \
            if (ATTR_HOMEKIT_DEVICE_CLASS in attributes) \
            else ATTR_DEVICE_CLASS

Commit:
Revise param/const/device_class_key based on advice

1. Remove state param;
2. Sort const;
3. Import ATTR_DEVICE_CLASS from homeassistant.const;
4. Remove device_class_key:
       device_class_key = ATTR_HOMEKIT_DEVICE_CLASS \
            if (ATTR_HOMEKIT_DEVICE_CLASS in attributes) \
            else ATTR_DEVICE_CLASS

Commit:
Wrap long line and add a blank line

Commit:
Wrap long line and add a blank line

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -2,13 +2,23 @@
+    DEVICE_CLASS_OCCUPANCY, SERV_OCCUPANCY_SENSOR, CHAR_OCCUPANCY_DETECTED,
+    DEVICE_CLASS_OPENING, SERV_CONTACT_SENSOR, CHAR_CONTACT_SENSOR_STATE,
+    DEVICE_CLASS_SMOKE, SERV_SMOKE_SENSOR, CHAR_SMOKE_DETECTED,
+    ATTR_HOMEKIT_DEVICE_CLASS)
```
`ATTR_HOMEKIT_DEVICE_CLASS` is an unused import.

Yonsm replied:
done

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +85,53 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+    DEVICE_CLASS_OPENING:
+        (SERV_CONTACT_SENSOR, CHAR_CONTACT_SENSOR_STATE),
+    DEVICE_CLASS_SMOKE:
+        (SERV_SMOKE_SENSOR, CHAR_SMOKE_DETECTED)}
```
Is that an option?
```py
BINARY_SENSOR_SERVICE_MAP = {
    DEVICE_CLASS_CO2: (SERV_CARBON_DIOXIDE_SENSOR,
                       CHAR_CARBON_DIOXIDE_DETECTED),
    DEVICE_CLASS_GAS: (SERV_CARBON_MONOXIDE_SENSOR,
                       CHAR_CARBON_MONOXIDE_DETECTED),
    DEVICE_CLASS_MOISTURE: (SERV_LEAK_SENSOR, CHAR_LEAK_DETECTED),
    DEVICE_CLASS_MOTION: (SERV_MOTION_SENSOR, CHAR_MOTION_DETECTED),
    DEVICE_CLASS_OCCUPANCY: (SERV_OCCUPANCY_SENSOR, CHAR_OCCUPANCY_DETECTED),
    DEVICE_CLASS_OPENING: (SERV_CONTACT_SENSOR, CHAR_CONTACT_SENSOR_STATE),
    DEVICE_CLASS_SMOKE: (SERV_SMOKE_SENSOR, CHAR_SMOKE_DETECTED)}
```

cdce8p replied:
Can you move the `BINARY_SENSOR_SERVICE_MAP` above the `TemperatureSensor` class. All constants should be defined belove the imports.

Yonsm replied:
done

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +85,53 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
                           self.entity_id, humidity)
+
+
+# Binary sensor service map
```
You don't need that.

Yonsm replied:
removed

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +85,53 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+        self.entity_id = entity_id
+
+        attributes = hass.states.get(entity_id).attributes
+        device_class = attributes.get(ATTR_DEVICE_CLASS)
```
You can combine those two lines and please use `self.hass` and `self.entity_id`:
```py
device_class = self.hass.states.get(self.entity_id).attributes.get(ATTR_DEVICE_CLASS)
```

Yonsm replied:
1.  Use attributes for homekit_device_class checking in future. But I'll follow your advice to remove it at this time.
2. I don't think self.hass is better then hass. Actually, local variable is better performance member variable, and it's good .

cdce8p replied:
It might be a little. However this will change with #13707. This PR will require to use self.hass and self.entity_id.

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +85,53 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+        attributes = hass.states.get(entity_id).attributes
+        device_class = attributes.get(ATTR_DEVICE_CLASS)
+        service_char = BINARY_SENSOR_SERVICE_MAP[device_class] \
+            if (device_class in BINARY_SENSOR_SERVICE_MAP) \
```
You don't need brackets here. `if device_class in BINARY_SENSOR_SERVICE_MAP \` is enough.

Commit:
Merge remote-tracking branch 'home-assistant/dev' into Yonsm_dev

# Conflicts:
#	homeassistant/components/homekit/const.py

Commit:
Merge remote-tracking branch 'home-assistant/dev' into Yonsm_dev

Commit:
Revise const and coding style

Commit:
Revise const and coding style

cdce8p committed:
Update Homekit to 1.1.9 (#13716)

* Version bump to HAP-python==1.1.9

* Updated types and tests

fabaff committed:
Update docstrings (#13720)

cdce8p commented on homeassistant/components/homekit/type_sensors.py:
```diff
@@ -75,3 +101,35 @@ def update_state(self, entity_id=None, old_state=None, new_state=None):
+
+        state = new_state.state
+        detected = (state == STATE_ON) or (state == STATE_HOME)
+        self.char_detected.set_value(detected, should_callback=False)
```
We just updated HAP-python to 1.1.9. 'Should_callback=False' will no longer be necessary. You probably need to rebase your branch, take take advantage of the update.

cdce8p reviewed (changes requested):
The BinarySensor class is looking good. Just on issue relate to an update of HAP-python. Can you add tests and the documentation? The tests should go in 'test_get_accessory' and 'test_type_sensor'.

cgtobi committed:
Add option to ignore availability in google calendar events (#13714)

bjw-s committed:
Allow use of date_string in service call (#13256)

* Allow use of date_string in service call

* Add stricter validation, fix descriptions

cdce8p committed:
Add timeout / debounce (for brightness and others) (#13534)

* Add async timeout feature

* Decorator for setter methods to limit service calls to HA
* Changed to async
* Use async_call_later
* Use lastargs, async_add_job

* Use dict for lastargs

* Updated tests to stop patch

hmn committed:
added support for smappee water sensors (#12831)

* added support for smappee water sensors

* fixed lint error and wrong location_id

* fixed lint error

* Use string formatting

dgomes committed:
Initialise filter_sensor with historical values (#13075)

* Initialise filter with historical values
Added get_last_state_changes()

* fix test

* Major changes to accommodate history + time_SMA

# Conflicts:
#	homeassistant/components/sensor/filter.py

* hail the hound!

* lint fixed

* less debug

* ups

* get state from the proper entity

* sensible default

* No defaults in get_last_state_changes

* list_reverseiterator instead of list

* prev_state to state

* Initialise filter with historical values
Added get_last_state_changes()

* fix test

* Major changes to accommodate history + time_SMA

# Conflicts:
#	homeassistant/components/sensor/filter.py

* hail the hound!

* lint fixed

* less debug

* ups

* get state from the proper entity

* sensible default

* No defaults in get_last_state_changes

* list_reverseiterator instead of list

* prev_state to state

* update

* added window_unit

* replace isinstance with window_unit

dlbroadfoot committed:
Fix Gogogate2 'available' attribute (#13728)

* Fixed bug -  unable to set base readaonly property

* PR fixes

* Added line

Commit:
Remove should_callback=False for HAP-python 1.1.9

And refine BINARY_SENSOR_SERVICE_MAP.

Commit:
Merge remote-tracking branch 'origin/Yonsm_dev' into Yonsm_dev

# Conflicts:
#	homeassistant/components/homekit/type_sensors.py

Yonsm commented:
Close this PR since it has more details. I'll re-PR later in a new branch.
